### ISSUE TITLE
Low perfomance on V100s

### ISSUE THREAD

**Issue by nidhidamodaran** [horovod/horovod#830]
I am trying to run horovod in a 2 node setup with 2 GPUs(V100) each. 

mpirun --allow-run-as-root --mca btl_openib_want_cuda_gdr 1  -x LD_LIBRARY_PATH -x PATH sh /examples/wrapper.sh 'python /examples/benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --variable_update=horovod  --model=resnet50 --device=gpu --distortions=False --init_learning_rate=.04 --optimizer=sgd --use_fp16=True --use_tf_layers=False --horovod_device=gpu --datasets_num_private_threads=4 --batch_size=64'

is the executed command,  The following are env variables:
 NCCL_IB_HCA=$IB_DEVICE
NCCL_SOCKET_IFNAME=net1
NCCL_IB_CUDA_SUPPORT=1
NCCL_IB_DISABLE=0
HOROVOD_HIERARCHICAL_ALLREDUCE=1
HOROVOD_FUSION_THRESHOLD=16777216
NCCL_MIN_NRINGS=4
OMPI_MCA_btl_openib_if_include==$IB_DEVICE

GPU topology:

              GPU0    GPU1
GPU0     X          SYS
GPU1    SYS       X

Performance per GPU is always in and around ~350. Is ther e any other parameter to tune so as to get the peek performance of 500 images/sec as claimed in #288

**Comment by alsrgv** [comment]
@nidhidamodaran, can you share the output of `nvidia-smi topo -m`, `ifconfig` and `ibv_devinfo -v` on both nodes?

**Comment by alsrgv** [comment]
Also, what images/sec do you get on a single GPU w/o Horovod?

**Comment by nidhidamodaran** [comment]
@alsrgv , without these 
NCCL_IB_CUDA_SUPPORT=1
NCCL_IB_DISABLE=0
HOROVOD_HIERARCHICAL_ALLREDUCE=1
HOROVOD_FUSION_THRESHOLD=16777216
NCCL_MIN_NRINGS=4

I am able to get 500 images/sec per GPU with horovod +RDMA , but if GPU direct is enabled, the  perf numbers drop to 110. What could be the issue here ?

**Comment by nidhidamodaran** [comment]
Logs with GPU direct
--------------------------
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO NET : Using interface net1:10.55.0.62<0>
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO NET/IB : Using interface net1 for sideband communication
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO NET/IB: [0] mlx5_18:1/RoCE 
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO Using internal Network IB
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO NET : Using interface net1:10.55.0.62<0>
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO NET/Socket : 1 interfaces found
NCCL version 2.2.13+cuda9.0
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO NET : Using interface net1:10.55.0.225<0>
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO NET/IB : Using interface net1 for sideband communication
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO NET/IB: [4] mlx5_5:1/RoCE 
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO Using internal Network IB
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO comm 0x7f51082a0320 rank 0 nranks 2
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO CUDA Dev 0, IB Ports : mlx5_18/1(SOC) 
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO comm 0x7f5c3c300ec0 rank 1 nranks 2
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO NET : Using interface net1:10.55.0.225<0>
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO NET/Socket : 1 interfaces found
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO CUDA Dev 0, IB Ports : mlx5_5/1(SOC) 
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO Using 256 threads
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO Min Comp Cap 7
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO NCCL_SINGLE_RING_THRESHOLD=262144
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO Ring 00 :    0   1
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO 0 -> 1 via NET/IB/0/GDRDMA
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO 1 -> 0 via NET/IB/0/GDRDMA
tensorflow-benchmarks-2-custom-worker-1:27162:27205 [0] INFO NET/IB: Dev 0 Port 1 qpn 17818 mtu 3 GID 0 (80FE/D80E73FEFFD81950)
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO NET/IB: Dev 0 Port 1 qpn 5630 mtu 3 GID 0 (80FE/A64D0FFEFFEE7A38)
tensorflow-benchmarks-2-custom-worker-0:29107:29150 [0] INFO Launch mode Parallel
Done warm up


Logs without GPU direct
------------------------------
ensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO NET : Using interface net1:10.55.0.62<0>
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO NET/IB : Using interface net1 for sideband communication
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO NET/IB: [0] mlx5_18:1/RoCE 
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO Using internal Network IB
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO NET : Using interface net1:10.55.0.62<0>
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO NET/Socket : 1 interfaces found
NCCL version 2.2.13+cuda9.0
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO NET : Using interface net1:10.55.0.225<0>
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO NET/IB : Using interface net1 for sideband communication
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO NET/IB: [4] mlx5_5:1/RoCE 
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO Using internal Network IB
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO comm 0x7f5c6c298b10 rank 0 nranks 2
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO CUDA Dev 0, IB Ports : mlx5_18/1(SOC) 
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO comm 0x7f22643014d0 rank 1 nranks 2
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO NET : Using interface net1:10.55.0.225<0>
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO NET/Socket : 1 interfaces found
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO CUDA Dev 0, IB Ports : mlx5_5/1(SOC) 
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO Using 256 threads
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO Min Comp Cap 7
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO NCCL_SINGLE_RING_THRESHOLD=262144
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO Ring 00 :    0   1
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO 0 -> 1 via NET/IB/0
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO 1 -> 0 via NET/IB/0
tensorflow-benchmarks-2-custom-worker-1:26586:26629 [0] INFO NET/IB: Dev 0 Port 1 qpn 16727 mtu 3 GID 0 (80FE/D80E73FEFFD81950)
tensorflow-benchmarks-2-custom-worker-0:28420:28463 [0] INFO NET/IB: Dev 0 Port 1 qpn 5627 mtu 3 GID 0 (80FE/A64D0FFEFFEE7A38)

**Comment by nidhidamodaran** [comment]
nvidia-smi topo -m
	GPU0	GPU1	mlx5_12	mlx5_25	mlx5_23	mlx5_10	mlx5_21	mlx5_8	mlx5_6	mlx5_19	CPU Affinity
GPU0	 X 	SYS	NODE	NODE	NODE	NODE	NODE	NODE	NODE	NODE	0-19
GPU1	SYS	 X 	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	20-39
mlx5_12	NODE	SYS	 X 	PIX	PIX	PIX	PIX	PIX	PIX	PIX	
mlx5_25	NODE	SYS	PIX	 X 	PIX	PIX	PIX	PIX	PIX	PIX	
mlx5_23	NODE	SYS	PIX	PIX	 X 	PIX	PIX	PIX	PIX	PIX	
mlx5_10	NODE	SYS	PIX	PIX	PIX	 X 	PIX	PIX	PIX	PIX	
mlx5_21	NODE	SYS	PIX	PIX	PIX	PIX	 X 	PIX	PIX	PIX	
mlx5_8	NODE	SYS	PIX	PIX	PIX	PIX	PIX	 X 	PIX	PIX	
mlx5_6	NODE	SYS	PIX	PIX	PIX	PIX	PIX	PIX	 X 	PIX	
mlx5_19	NODE	SYS	PIX	PIX	PIX	PIX	PIX	PIX	PIX	 X

**Comment by nidhidamodaran** [comment]
with 2 workers, horovod + RDMA, i get 560 images/s on each worker
with 2 workers, horovod +GDRDMA, i get 116 images/sec on each worker

**Comment by alsrgv** [comment]
@nidhidamodaran, do you have [nv_peer_memory](https://github.com/Mellanox/nv_peer_memory) driver installed & running on both machines?  If yes, you may need to Mellanox to help with debugging.

**Comment by nidhidamodaran** [comment]
I have another weird issue with scaling across node. After the number of workers increases beyond 3, Images/s per worker drops down to 20 Images/sec.

After setting these  I get  around 200 Images/sec per worker. But till 3 workers, i get 500 Images/sec. This is 3 node setup

HOROVOD_HIERARCHICAL_ALLREDUCE=1
HOROVOD_FUSION_THRESHOLD=16777216
NCCL_MIN_NRINGS=4

I am at loss understanding what could be going wrong.
